### PR TITLE
nixos/switch-to-configuration: Mount improvements and a lot more test cases

### LINKED ISSUE #242737: switch-to-configuration tries to stop .mount units and fails
### Describe the bug
The `switch-to-configuration` script is apparently trying to stop and start
changed .mount units, presumably to cause a `remount`?
If that happens the script fails and the new configuration is not activated.
Worse, the system is left in a broken state with some of the stopped (as part
of the switch) units never restarted.

### Steps To Reproduce
1. change the value of `fileSystems."/".fsType`, for example from `ext4` to `auto`.
2. run `nixos-rebuild test`

### Expected behavior
The unit should probably not be stopped and the switch should succeed.

### Screenshots
```
# nixos-rebuild switch
stopping the following units: -.mount
warning: error(s) occurred while switching to the new configuration
```
If using `--target-host` remotely also the SSH connection drops.

### Additional context

NixOS version: 23.05.1194.b6c73c5fe53 (Stoat)

The filesystem entry is:
```
  fileSystems."/" =
    { device = "/dev/main/nixos";
      fsType = "ext4";
    };
```

### Notify maintainers

@dasJ 

cc: @Artturin

### PR DESCRIPTION
**Pull request by dasJ**
## Description of changes

Adds way more test cases, mostly about command line handling and fstab handling.
Also fixes #242737 by adding an explicit exclusion as a double safeguard.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by stigtsp — APPROVED** [review]
Code LGTM

**Review by oddlama — APPROVED** [review]
LGTM

File: nixos/doc/manual/development/what-happens-during-a-system-switch.chapter.md [hunk 1]
```diff
@@ -21,8 +21,9 @@ If the action is `switch` or `test`, the currently running system is inspected
 and the actions to switch to the new system are calculated. This process takes
 two data sources into account: `/etc/fstab` and the current systemd status.
 Mounts and swaps are read from `/etc/fstab` and the corresponding actions are
-generated. If a new mount is added, for example, the proper `.mount` unit is
-marked to be started. The current systemd state is inspected, the difference
+generated. If a options of a mount are modified, for example, the proper `.mount`
```
Comment (rnhmjoj) [inline nixos/doc/manual/development/what-happens-during-a-system-switch.chapter.md:24]:
```suggestion
generated. If the options of a mount are modified, for example, the proper `.mount`
```

**Comment by stigtsp** [comment]
@ofborg test switchTest

**Pull request merged**